src/client/coll/results.rs

```diff
@@ -87,39 +87,81 @@ impl BulkWriteResult {
-        match result.write_exception {
-            Some(write_exception) =>
-                exception.add_write_exception(write_exception, i as i32, req),
-            None => {
-                let id = result.inserted_id.expect("`inserted_id` should not be `None` \
-                                                    if there is no WriteException");
-                self.inserted_ids.insert(i, id);
-                self.inserted_count += 1;
+    pub fn process_bulk_delete_result(&mut self, result: BulkDeleteResult,
+                                      models: Vec<WriteModel>,
+                                      exception: &mut BulkWriteException) -> bool {
+        let ok = exception.add_bulk_write_exception(result.write_exception, models);
+        self.deleted_count += result.deleted_count;
 
-                exception.processed_requests.push(req)
-            }
-        };
+        ok
     }
 
     pub fn process_insert_many_result(&mut self, result: InsertManyResult,
-                                      models: Vec<WriteModel>,
-                                      exception: &mut BulkWriteException) {
-        match result.bulk_write_exception {
-            Some(new_exception) =>
-                exception.add_bulk_write_exception(new_exception),
-            None => for model in models {
-                exception.processed_requests.push(model);
-            }
-        }
+                                      models: Vec<WriteModel>, start_index: i64,
+                                      exception: &mut BulkWriteException) -> bool {
+        let ok = exception.add_bulk_write_exception(result.bulk_write_exception, models);
 
         if let Some(ids) = result.inserted_ids {
             for (i, id) in ids {
-                self.inserted_ids.insert(i, id);
+                self.inserted_ids.insert(start_index + i, id);
+                self.inserted_count += 1;
             }
         }
+
+        ok
+    }
+
+    fn parse_upserted_id(mut document: bson::Document, start_index: i64,
+                         upserted_ids: &mut BTreeMap<i64, Bson>) -> i32 {
+        let (index, id) = (document.remove("index"), document.remove("_id"));
+
+        match (index, id) {
+            (Some(Bson::I32(i)), Some(bson_id)) => {
+                let _ = upserted_ids.insert(start_index + i as i64, bson_id);
+                1
+            }
+            (Some(Bson::I64(i)), Some(bson_id)) => {
+                let _ = upserted_ids.insert(start_index + i, bson_id.clone());
+                1
+            }
+            _ => 0
+        }
+    }
+
+    fn parse_upserted_ids(bson: Bson, start_index: i64,
+                          upserted_ids: &mut BTreeMap<i64, Bson>) -> i32 {
+        match bson {
+            Bson::Document(doc) => BulkWriteResult::parse_upserted_id(doc, start_index, upserted_ids),
+            Bson::Array(vec) => {
+                let mut count = 0;
+
+                for bson in vec {
```

Comment:
Not really sure what else to call it, given that it's some arbitrary bson value; there's no way to know if it's a document, array, number, or anything else.